# Task
Address the data science challenge using the data from "challenge1-data-science-latam-main.zip".

## Carga y Exploración de Datos

### Subtask:
Cargar los datos del challenge y realizar una exploración inicial para entender su estructura, tipos de variables, valores faltantes y estadísticas descriptivas.


**Reasoning**:
The first step is to decompress the provided zip file to access the data, which is a necessary prerequisite for loading the data into a pandas DataFrame.



In [1]:
import zipfile
import os

zip_file_path = '/content/challenge1-data-science-latam-main.zip'
extraction_path = '/content/challenge_data'

# Create the extraction directory if it doesn't exist
os.makedirs(extraction_path, exist_ok=True)

# Unzip the file
with zipfile.ZipFile(zip_file_path, 'r') as zip_ref:
    zip_ref.extractall(extraction_path)

print(f"File '{zip_file_path}' unzipped to '{extraction_path}' successfully.")

# List the contents of the extracted directory to identify the main data file
print("Contents of the extracted directory:")
for root, dirs, files in os.walk(extraction_path):
    for name in files:
        print(os.path.join(root, name))
    for name in dirs:
        print(os.path.join(root, name))

File '/content/challenge1-data-science-latam-main.zip' unzipped to '/content/challenge_data' successfully.
Contents of the extracted directory:
/content/challenge_data/challenge1-data-science-latam-main
/content/challenge_data/challenge1-data-science-latam-main/AluraStoreLatam.ipynb
/content/challenge_data/challenge1-data-science-latam-main/base-de-datos-challenge1-latam
/content/challenge_data/challenge1-data-science-latam-main/base-de-datos-challenge1-latam/tienda_3.csv
/content/challenge_data/challenge1-data-science-latam-main/base-de-datos-challenge1-latam/tienda_2.csv
/content/challenge_data/challenge1-data-science-latam-main/base-de-datos-challenge1-latam/tienda_4.csv
/content/challenge_data/challenge1-data-science-latam-main/base-de-datos-challenge1-latam/tienda_1 .csv


**Reasoning**:
The previous step successfully unzipped the archive and listed the CSV files. Now, I need to load these multiple CSV files into pandas DataFrames and concatenate them into a single DataFrame to fulfill the instruction of 'Carga el archivo de datos principal en un DataFrame de pandas'.



In [2]:
import pandas as pd
import os

data_directory = '/content/challenge_data/challenge1-data-science-latam-main/base-de-datos-challenge1-latam'

# List all CSV files in the directory
csv_files = [f for f in os.listdir(data_directory) if f.endswith('.csv')]

# Create a list to store individual DataFrames
dfs = []

# Read each CSV file and append its DataFrame to the list
for file in csv_files:
    file_path = os.path.join(data_directory, file)
    # Some file names have leading/trailing spaces, so strip them before reading
    try:
        df_temp = pd.read_csv(file_path.strip())
        dfs.append(df_temp)
        print(f"Loaded {file} with {len(df_temp)} rows.")
    except Exception as e:
        print(f"Error loading {file}: {e}")

# Concatenate all DataFrames into a single one
if dfs:
    df = pd.concat(dfs, ignore_index=True)
    print(f"\nSuccessfully concatenated {len(dfs)} files into a single DataFrame with {len(df)} rows and {len(df.columns)} columns.")
else:
    df = pd.DataFrame()
    print("No CSV files found or loaded.")

# Display the first few rows of the combined DataFrame
print("\nFirst 5 rows of the combined DataFrame:")
display(df.head())

Loaded tienda_3.csv with 2359 rows.
Loaded tienda_2.csv with 2359 rows.
Loaded tienda_4.csv with 2358 rows.
Loaded tienda_1 .csv with 2359 rows.

Successfully concatenated 4 files into a single DataFrame with 9435 rows and 12 columns.

First 5 rows of the combined DataFrame:


,Producto,Categoría del Producto,Precio,Costo de envío,Fecha de Compra,Vendedor,Lugar de Compra,Calificación,Método de pago,Cantidad de cuotas,lat,lon
0,Secadora de ropa,Electrodomésticos,1285100.0,66600.0,17/12/2021,Camila Rivera,Santa Marta,4,Tarjeta de crédito,4,11.24079,-74.19904
1,Mesa de noche,Muebles,290100.0,13600.0,10/02/2021,Maria Alfonso,Cali,3,Tarjeta de crédito,1,3.43722,-76.52250
2,Guitarra acústica,Instrumentos musicales,303500.0,17400.0,22/01/2023,Lucas Olivera,Bogotá,5,Tarjeta de crédito,3,4.60971,-74.08175
3,Balón de voleibol,Deportes y diversión,58800.0,3300.0,22/03/2022,Pedro Gomez,Barranquilla,4,Tarjeta de crédito,6,10.96854,-74.78132
4,Armario,Muebles,521100.0,25900.0,12/01/2022,Blanca Ramirez,Bogotá,1,Tarjeta de crédito,1,4.60971,-74.08175


**Reasoning**:
To understand the structure of the DataFrame, I will display general information including column data types and non-null counts, as requested by the subtask.



In [3]:
print('DataFrame Info:')
df.info()

DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9435 entries, 0 to 9434
Data columns (total 12 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Producto                9435 non-null   object 
 1   Categoría del Producto  9435 non-null   object 
 2   Precio                  9435 non-null   float64
 3   Costo de envío          9435 non-null   float64
 4   Fecha de Compra         9435 non-null   object 
 5   Vendedor                9435 non-null   object 
 6   Lugar de Compra         9435 non-null   object 
 7   Calificación            9435 non-null   int64  
 8   Método de pago          9435 non-null   object 
 9   Cantidad de cuotas      9435 non-null   int64  
 10  lat                     9435 non-null   float64
 11  lon                     9435 non-null   float64
dtypes: float64(4), int64(2), object(6)
memory usage: 884.7+ KB


**Reasoning**:
To directly address the subtask's instruction to calculate and display the number of missing values per column, I will use `df.isnull().sum()`.



In [4]:
print('Missing values per column:')
print(df.isnull().sum())

Missing values per column:
Producto                  0
Categoría del Producto    0
Precio                    0
Costo de envío            0
Fecha de Compra           0
Vendedor                  0
Lugar de Compra           0
Calificación              0
Método de pago            0
Cantidad de cuotas        0
lat                       0
lon                       0
dtype: int64


**Reasoning**:
To complete the subtask, I need to generate descriptive statistics for the numerical columns of the DataFrame, as specified in the instructions.



In [5]:
print('Descriptive statistics for numerical columns:')
print(df.describe())

Descriptive statistics for numerical columns:
             Precio  Costo de envío  Calificación  Cantidad de cuotas  \
count  9.435000e+03     9435.000000   9435.000000         9435.000000   
mean   4.667323e+05    24875.145734      4.014520            2.927398   
std    5.903207e+05    31480.036346      1.390518            2.802776   
min    7.600000e+03        0.000000      1.000000            1.000000   
25%    5.520000e+04     3000.000000      3.000000            1.000000   
50%    2.215000e+05    11700.000000      5.000000            1.000000   
75%    6.633000e+05    34900.000000      5.000000            4.000000   
max    2.977000e+06   162500.000000      5.000000           24.000000   

               lat          lon  
count  9435.000000  9435.000000  
mean      5.428909   -74.760108  
std       2.355663     1.207155  
min      -4.215280   -77.281110  
25%       4.609710   -75.563590  
50%       4.609710   -74.199040  
75%       6.251840   -74.081750  
max      11.544440   -67

## Preprocesamiento de Datos

### Subtask:
Limpiar y transformar los datos, incluyendo la ingeniería de características de la columna de fecha, la codificación de variables categóricas y el escalado de características numéricas.


**Reasoning**:
The first step in the data preprocessing subtask is to convert the 'Fecha de Compra' column to a datetime object, which is essential for extracting temporal features and performing date-based calculations as per the instructions.



In [6]:
df['Fecha de Compra'] = pd.to_datetime(df['Fecha de Compra'])
print("Converted 'Fecha de Compra' to datetime format.")
print(df['Fecha de Compra'].head())

Converted 'Fecha de Compra' to datetime format.
0   2021-12-17
1   2021-02-10
2   2023-01-22
3   2022-03-22
4   2022-01-12
Name: Fecha de Compra, dtype: datetime64[ns]


/tmp/ipython-input-390/3811415767.py:1: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df['Fecha de Compra'] = pd.to_datetime(df['Fecha de Compra'])


**Reasoning**:
The previous `pd.to_datetime` conversion issued a `UserWarning` due to ambiguous date format. To address this and ensure correct parsing, I will explicitly set `dayfirst=True` to handle the `DD/MM/YYYY` format and silence the warning.



In [7]:
df['Fecha de Compra'] = pd.to_datetime(df['Fecha de Compra'], dayfirst=True)
print("Converted 'Fecha de Compra' to datetime format, handling dayfirst parsing.")
print(df['Fecha de Compra'].head())

Converted 'Fecha de Compra' to datetime format, handling dayfirst parsing.
0   2021-12-17
1   2021-02-10
2   2023-01-22
3   2022-03-22
4   2022-01-12
Name: Fecha de Compra, dtype: datetime64[ns]


**Reasoning**:
Now that 'Fecha de Compra' is in datetime format, I will extract temporal features like year, month, day, day of the week, and day of the year, as specified in the instructions, and store them as new columns.



In [8]:
df['Año_Compra'] = df['Fecha de Compra'].dt.year
df['Mes_Compra'] = df['Fecha de Compra'].dt.month
df['Dia_Compra'] = df['Fecha de Compra'].dt.day
df['Dia_Semana_Compra'] = df['Fecha de Compra'].dt.dayofweek # Monday=0, Sunday=6
df['Dia_Año_Compra'] = df['Fecha de Compra'].dt.dayofyear

print("Extracted temporal features from 'Fecha de Compra':")
print(df[['Fecha de Compra', 'Año_Compra', 'Mes_Compra', 'Dia_Compra', 'Dia_Semana_Compra', 'Dia_Año_Compra']].head())

Extracted temporal features from 'Fecha de Compra':
  Fecha de Compra  Año_Compra  Mes_Compra  Dia_Compra  Dia_Semana_Compra  \
0      2021-12-17        2021          12          17                  4   
1      2021-02-10        2021           2          10                  2   
2      2023-01-22        2023           1          22                  6   
3      2022-03-22        2022           3          22                  1   
4      2022-01-12        2022           1          12                  2   

   Dia_Año_Compra  
0             351  
1              41  
2              22  
3              81  
4              12  


**Reasoning**:
To calculate the difference in days from the earliest purchase, I will first find the minimum 'Fecha de Compra' and then subtract it from each 'Fecha de Compra', converting the timedelta to days, as specified in the instructions.



In [9]:
fecha_minima = df['Fecha de Compra'].min()
df['Dias_Desde_Primera_Compra'] = (df['Fecha de Compra'] - fecha_minima).dt.days

print("Calculated 'Dias_Desde_Primera_Compra' column:")
print(df[['Fecha de Compra', 'Dias_Desde_Primera_Compra']].head())

Calculated 'Dias_Desde_Primera_Compra' column:
  Fecha de Compra  Dias_Desde_Primera_Compra
0      2021-12-17                        716
1      2021-02-10                        406
2      2023-01-22                       1117
3      2022-03-22                        811
4      2022-01-12                        742


**Reasoning**:
Now that temporal features have been extracted and a time difference column created, I will identify the remaining nominal categorical columns and apply one-hot encoding, as specified by the subtask instructions. This prepares the data for modeling by converting categorical text data into a numerical format.



In [10]:
categorical_cols = ['Producto', 'Categoría del Producto', 'Vendedor', 'Lugar de Compra', 'Método de pago']

# Apply one-hot encoding
df = pd.get_dummies(df, columns=categorical_cols, drop_first=True)

print("One-hot encoded categorical columns. New DataFrame shape:")
print(df.shape)
print("First 5 rows of the DataFrame with one-hot encoded columns:")
display(df.head())

One-hot encoded categorical columns. New DataFrame shape:
(9435, 104)
First 5 rows of the DataFrame with one-hot encoded columns:


,Precio,Costo de envío,Fecha de Compra,Calificación,Cantidad de cuotas,lat,lon,Año_Compra,Mes_Compra,Dia_Compra,...,Lugar de Compra_Pasto,Lugar de Compra_Pereira,Lugar de Compra_Riohacha,Lugar de Compra_Santa Marta,Lugar de Compra_Soacha,Lugar de Compra_Valledupar,Lugar de Compra_Villavicencio,Método de pago_Nequi,Método de pago_Tarjeta de crédito,Método de pago_Tarjeta de débito
0,1285100.0,66600.0,2021-12-17,4,4,11.24079,-74.19904,2021,12,17,...,False,False,False,True,False,False,False,False,True,False
1,290100.0,13600.0,2021-02-10,3,1,3.43722,-76.52250,2021,2,10,...,False,False,False,False,False,False,False,False,True,False
2,303500.0,17400.0,2023-01-22,5,3,4.60971,-74.08175,2023,1,22,...,False,False,False,False,False,False,False,False,True,False
3,58800.0,3300.0,2022-03-22,4,6,10.96854,-74.78132,2022,3,22,...,False,False,False,False,False,False,False,False,True,False
4,521100.0,25900.0,2022-01-12,1,1,4.60971,-74.08175,2022,1,12,...,False,False,False,False,False,False,False,False,True,False


**Reasoning**:
After extracting temporal features and one-hot encoding categorical variables, I will now drop the original 'Fecha de Compra' column, as it has been transformed into multiple numerical features.



In [11]:
df = df.drop('Fecha de Compra', axis=1)
print("Dropped original 'Fecha de Compra' column.")
print(f"New DataFrame shape: {df.shape}")
print("First 5 rows of the DataFrame after dropping 'Fecha de Compra':")
display(df.head())

Dropped original 'Fecha de Compra' column.
New DataFrame shape: (9435, 103)
First 5 rows of the DataFrame after dropping 'Fecha de Compra':


,Precio,Costo de envío,Calificación,Cantidad de cuotas,lat,lon,Año_Compra,Mes_Compra,Dia_Compra,Dia_Semana_Compra,...,Lugar de Compra_Pasto,Lugar de Compra_Pereira,Lugar de Compra_Riohacha,Lugar de Compra_Santa Marta,Lugar de Compra_Soacha,Lugar de Compra_Valledupar,Lugar de Compra_Villavicencio,Método de pago_Nequi,Método de pago_Tarjeta de crédito,Método de pago_Tarjeta de débito
0,1285100.0,66600.0,4,4,11.24079,-74.19904,2021,12,17,4,...,False,False,False,True,False,False,False,False,True,False
1,290100.0,13600.0,3,1,3.43722,-76.52250,2021,2,10,2,...,False,False,False,False,False,False,False,False,True,False
2,303500.0,17400.0,5,3,4.60971,-74.08175,2023,1,22,6,...,False,False,False,False,False,False,False,False,True,False
3,58800.0,3300.0,4,6,10.96854,-74.78132,2022,3,22,1,...,False,False,False,False,False,False,False,False,True,False
4,521100.0,25900.0,1,1,4.60971,-74.08175,2022,1,12,2,...,False,False,False,False,False,False,False,False,True,False


**Reasoning**:
To complete the preprocessing subtask, I will identify the numerical columns that need scaling, as specified in the instructions. I will then apply `StandardScaler` to these columns to standardize their values, which is beneficial for many machine learning algorithms.



In [12]:
from sklearn.preprocessing import StandardScaler

# Identify numerical columns to scale
# Exclude one-hot encoded columns (which are boolean/binary) and the newly created date features that are already good to go.
# 'Calificación' and 'Cantidad de cuotas' are numerical but might be better treated as categorical if there are few unique values,
# however, the instruction explicitly says to scale numerical features like 'Precio' and 'Costo de envío', and these are also numerical.
# Let's scale all continuous numerical features and the extracted date features for consistency.

# Identify all numerical columns (excluding the one-hot encoded boolean columns)
# Get all column names that are not boolean (one-hot encoded columns are boolean after get_dummies)
numerical_cols_to_scale = df.select_dtypes(include=['number']).columns.tolist()

# Filter out potential one-hot encoded columns if they were identified as numbers, which they are (0s and 1s).
# We want to scale continuous numerical features and temporal features, not binary flags.
# A simple heuristic is to exclude columns with only 0 and 1 values, unless they were originally intended as such.
# Given 'Precio', 'Costo de envío', 'Calificación', 'Cantidad de cuotas', 'lat', 'lon' and the date-derived features,
# we should explicitly select these as they are the primary continuous numerical features.

# Let's define the numerical columns explicitly based on the original data and the derived date features
explicit_numerical_cols = [
    'Precio', 'Costo de envío', 'Calificación', 'Cantidad de cuotas', 'lat', 'lon',
    'Año_Compra', 'Mes_Compra', 'Dia_Compra', 'Dia_Semana_Compra', 'Dia_Año_Compra', 'Dias_Desde_Primera_Compra'
]

# Ensure all explicit_numerical_cols actually exist in df columns before proceeding
explicit_numerical_cols = [col for col in explicit_numerical_cols if col in df.columns]

scaler = StandardScaler()
df[explicit_numerical_cols] = scaler.fit_transform(df[explicit_numerical_cols])

print("Scaled numerical columns using StandardScaler.")
print("Descriptive statistics for scaled numerical columns:")
print(df[explicit_numerical_cols].describe())
print("First 5 rows of the DataFrame with scaled numerical columns:")
display(df.head())

Scaled numerical columns using StandardScaler.
Descriptive statistics for scaled numerical columns:
             Precio  Costo de envío  Calificación  Cantidad de cuotas  \
count  9.435000e+03    9.435000e+03  9.435000e+03        9.435000e+03   
mean   2.409896e-17   -9.790202e-18  3.144161e-16        5.158683e-17   
std    1.000053e+00    1.000053e+00  1.000053e+00        1.000053e+00   
min   -7.778088e-01   -7.902298e-01 -2.168027e+00       -6.877112e-01   
25%   -6.971704e-01   -6.949263e-01 -7.296377e-01       -6.877112e-01   
50%   -4.154442e-01   -4.185460e-01  7.087517e-01       -6.877112e-01   
75%    3.330023e-01    3.184680e-01  7.087517e-01        3.827131e-01   
max    4.252605e+00    4.372045e+00  7.087517e-01        7.518875e+00   

               lat           lon    Año_Compra    Mes_Compra    Dia_Compra  \
count  9435.000000  9.435000e+03  9.435000e+03  9.435000e+03  9.435000e+03   
mean      0.000000  6.581275e-15 -5.379603e-14 -1.430876e-17 -5.648193e-17   
std     

,Precio,Costo de envío,Calificación,Cantidad de cuotas,lat,lon,Año_Compra,Mes_Compra,Dia_Compra,Dia_Semana_Compra,...,Lugar de Compra_Pasto,Lugar de Compra_Pereira,Lugar de Compra_Riohacha,Lugar de Compra_Santa Marta,Lugar de Compra_Soacha,Lugar de Compra_Valledupar,Lugar de Compra_Villavicencio,Método de pago_Nequi,Método de pago_Tarjeta de crédito,Método de pago_Tarjeta de débito
0,1.386384,1.325509,-0.010443,0.382713,2.467325,0.464810,-0.163001,1.637832,0.133939,0.487506,...,False,False,False,True,False,False,False,False,True,False
1,-0.299230,-0.358187,-0.729638,-0.687711,-0.845535,-1.460032,-0.163001,-1.170875,-0.666099,-0.516558,...,False,False,False,False,False,False,False,False,True,False
2,-0.276529,-0.237469,0.708752,0.025905,-0.347776,0.561978,1.958257,-1.451746,0.705396,1.491570,...,False,False,False,False,False,False,False,False,True,False
3,-0.691072,-0.685396,-0.010443,1.096329,2.351747,-0.017572,0.897628,-0.890004,0.705396,-1.018590,...,False,False,False,False,False,False,False,False,True,False
4,0.092103,0.032557,-2.168027,-0.687711,-0.347776,0.561978,0.897628,-1.451746,-0.437517,-0.516558,...,False,False,False,False,False,False,False,False,True,False


## Construcción y Entrenamiento del Modelo

### Subtask:
Seleccionar y entrenar un modelo de machine learning adecuado para la tarea del challenge.


**Reasoning**:
To begin the model building process, I need to define the target variable 'y' as 'Calificación' and the features 'X' as all other columns from the preprocessed DataFrame `df`, as specified by the instructions.



In [13]:
y = df['Calificación']
X = df.drop('Calificación', axis=1)

print(f"Shape of X: {X.shape}")
print(f"Shape of y: {y.shape}")
print("First 5 rows of X:")
display(X.head())
print("First 5 values of y:")
display(y.head())

Shape of X: (9435, 102)
Shape of y: (9435,)
First 5 rows of X:


,Precio,Costo de envío,Cantidad de cuotas,lat,lon,Año_Compra,Mes_Compra,Dia_Compra,Dia_Semana_Compra,Dia_Año_Compra,...,Lugar de Compra_Pasto,Lugar de Compra_Pereira,Lugar de Compra_Riohacha,Lugar de Compra_Santa Marta,Lugar de Compra_Soacha,Lugar de Compra_Valledupar,Lugar de Compra_Villavicencio,Método de pago_Nequi,Método de pago_Tarjeta de crédito,Método de pago_Tarjeta de débito
0,1.386384,1.325509,0.382713,2.467325,0.464810,-0.163001,1.637832,0.133939,0.487506,1.640426,...,False,False,False,True,False,False,False,False,True,False
1,-0.299230,-0.358187,-0.687711,-0.845535,-1.460032,-0.163001,-1.170875,-0.666099,-0.516558,-1.208910,...,False,False,False,False,False,False,False,False,True,False
2,-0.276529,-0.237469,0.025905,-0.347776,0.561978,1.958257,-1.451746,0.705396,1.491570,-1.383547,...,False,False,False,False,False,False,False,False,True,False
3,-0.691072,-0.685396,1.096329,2.351747,-0.017572,0.897628,-0.890004,0.705396,-1.018590,-0.841254,...,False,False,False,False,False,False,False,False,True,False
4,0.092103,0.032557,-0.687711,-0.347776,0.561978,0.897628,-1.451746,-0.437517,-0.516558,-1.475461,...,False,False,False,False,False,False,False,False,True,False


First 5 values of y:


,Calificación
0,-0.010443
1,-0.729638
2,0.708752
3,-0.010443
4,-2.168027


**Reasoning**:
To prepare the data for model training and evaluation, I will split the `X` (features) and `y` (target) data into training and testing sets, as specified in the instructions, using a 20% test size and a `random_state` for reproducibility.



In [14]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(f"Shape of X_train: {X_train.shape}")
print(f"Shape of X_test: {X_test.shape}")
print(f"Shape of y_train: {y_train.shape}")
print(f"Shape of y_test: {y_test.shape}")

Shape of X_train: (7548, 102)
Shape of X_test: (1887, 102)
Shape of y_train: (7548,)
Shape of y_test: (1887,)


**Reasoning**:
Now that the data is split into training and testing sets, I will import `RandomForestClassifier` and instantiate it, setting a `random_state` for reproducibility, and then train the model using the training data, as per the instructions.



In [15]:
from sklearn.ensemble import RandomForestClassifier

# Instantiate the RandomForestClassifier
# Setting random_state for reproducibility
model = RandomForestClassifier(random_state=42)

# Train the model
model.fit(X_train, y_train)

print("RandomForestClassifier model trained successfully.")

ValueError: Unknown label type: continuous. Maybe you are trying to fit a classifier, which expects discrete classes on a regression target with continuous values.

**Reasoning**:
The previous step failed because the `RandomForestClassifier` received a continuous target variable (`y_train`) due to earlier scaling of 'Calificación'. To fix this, I need to inverse transform and discretize the `y_train` and `y_test` back to their original integer ratings (1-5) before training the classifier. This ensures the target variable is in the expected discrete format for classification.



In [16]:
from sklearn.ensemble import RandomForestClassifier
import numpy as np
import pandas as pd

# Original mean and standard deviation for 'Calificación' before scaling
# These values were obtained from the output of df.describe() in a previous step.
original_calificacion_mean = 4.014520
original_calificacion_std = 1.390518

# Inverse transform y_train and y_test to get approximate original integer values
y_train_unscaled = (y_train * original_calificacion_std) + original_calificacion_mean
y_test_unscaled = (y_test * original_calificacion_std) + original_calificacion_mean

# Convert to integer type, which is required for classification
# Rounding is important because the original values were integers (1-5)
y_train_discrete = np.round(y_train_unscaled).astype(int)
y_test_discrete = np.round(y_test_unscaled).astype(int)

# Ensure the discrete values are within the valid range (1-5) of the original ratings
y_train_discrete = np.clip(y_train_discrete, 1, 5)
y_test_discrete = np.clip(y_test_discrete, 1, 5)


# Instantiate the RandomForestClassifier
# Setting random_state for reproducibility
model = RandomForestClassifier(random_state=42)

# Train the model with the discrete target variable
model.fit(X_train, y_train_discrete)

print("RandomForestClassifier model trained successfully with discrete 'Calificación' values.")

# Optional: Print value counts to verify discrete target
print("\nValue counts for y_train_discrete:")
print(pd.Series(y_train_discrete).value_counts())
print("\nValue counts for y_test_discrete:")
print(pd.Series(y_test_discrete).value_counts())

RandomForestClassifier model trained successfully with discrete 'Calificación' values.

Value counts for y_train_discrete:
Calificación
5    4178
4    1437
1     987
3     673
2     273
Name: count, dtype: int64

Value counts for y_test_discrete:
Calificación
5    1072
4     376
1     212
3     157
2      70
Name: count, dtype: int64


## Evaluación del Modelo

### Subtask:
Evaluar el rendimiento del modelo utilizando métricas apropiadas para el problema (ej. precisión, recall, F1-score, RMSE, MAE). Optimización de hiperparámetros si es necesario.


**Reasoning**:
To evaluate the trained model, I will first make predictions on the test set (`X_test`) using the `model.predict()` method and store the results in `y_pred`.



In [17]:
y_pred = model.predict(X_test)

print("Model predictions on X_test generated.")
print("First 5 predictions:")
print(y_pred[:5])

Model predictions on X_test generated.
First 5 predictions:
[5 5 5 5 5]


**Reasoning**:
Now that predictions have been generated, I will calculate the accuracy score and a classification report using `y_test_discrete` and `y_pred` to evaluate the model's performance, as specified in the instructions.



In [18]:
from sklearn.metrics import accuracy_score, classification_report

# Calculate accuracy
accuracy = accuracy_score(y_test_discrete, y_pred)
print(f"\nModel Accuracy: {accuracy:.4f}")

# Generate classification report
report = classification_report(y_test_discrete, y_pred)
print("\nClassification Report:\n", report)


Model Accuracy: 0.5601

Classification Report:
               precision    recall  f1-score   support

           1       0.00      0.00      0.00       212
           2       0.00      0.00      0.00        70
           3       0.33      0.01      0.01       157
           4       0.24      0.03      0.05       376
           5       0.57      0.98      0.72      1072

    accuracy                           0.56      1887
   macro avg       0.23      0.20      0.16      1887
weighted avg       0.40      0.56      0.42      1887



/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


**Reasoning**:
The previous classification report showed `UndefinedMetricWarning` and very low precision/recall for minority classes, indicating a class imbalance problem. To address this, I will re-instantiate and retrain the `RandomForestClassifier` with `class_weight='balanced'` to automatically adjust weights inversely proportional to class frequencies, aiming to improve the model's ability to predict all classes.



In [19]:
from sklearn.ensemble import RandomForestClassifier
import numpy as np
import pandas as pd

# Original mean and standard deviation for 'Calificación' before scaling
# These values were obtained from the output of df.describe() in a previous step.
original_calificacion_mean = 4.014520
original_calificacion_std = 1.390518

# Inverse transform y_train and y_test to get approximate original integer values
y_train_unscaled = (y_train * original_calificacion_std) + original_calificacion_mean
y_test_unscaled = (y_test * original_calificacion_std) + original_calificacion_mean

# Convert to integer type, which is required for classification
# Rounding is important because the original values were integers (1-5)
y_train_discrete = np.round(y_train_unscaled).astype(int)
y_test_discrete = np.round(y_test_unscaled).astype(int)

# Ensure the discrete values are within the valid range (1-5) of the original ratings
y_train_discrete = np.clip(y_train_discrete, 1, 5)
y_test_discrete = np.clip(y_test_discrete, 1, 5)


# Instantiate the RandomForestClassifier with class_weight='balanced'
# Setting random_state for reproducibility
model = RandomForestClassifier(random_state=42, class_weight='balanced')

# Train the model with the discrete target variable
model.fit(X_train, y_train_discrete)

print("RandomForestClassifier model trained successfully with discrete 'Calificación' values and balanced class weights.")

# Optional: Print value counts to verify discrete target
print("\nValue counts for y_train_discrete:")
print(pd.Series(y_train_discrete).value_counts())
print("\nValue counts for y_test_discrete:")
print(pd.Series(y_test_discrete).value_counts())

RandomForestClassifier model trained successfully with discrete 'Calificación' values and balanced class weights.

Value counts for y_train_discrete:
Calificación
5    4178
4    1437
1     987
3     673
2     273
Name: count, dtype: int64

Value counts for y_test_discrete:
Calificación
5    1072
4     376
1     212
3     157
2      70
Name: count, dtype: int64


**Reasoning**:
After retraining the model with `class_weight='balanced'`, I will now make new predictions on the test set (`X_test`) using the updated model to evaluate its performance against the discrete target variable.



In [20]:
y_pred = model.predict(X_test)

print("Model predictions on X_test generated with balanced class weights.")
print("First 5 predictions:")
print(y_pred[:5])

Model predictions on X_test generated with balanced class weights.
First 5 predictions:
[5 5 5 5 5]


**Reasoning**:
Now that the model has been retrained with balanced class weights and new predictions (`y_pred`) have been generated, I will calculate the accuracy score and a classification report using the `y_test_discrete` and the new `y_pred` to assess the improved performance of the model, as specified in the instructions.



In [21]:
from sklearn.metrics import accuracy_score, classification_report

# Calculate accuracy with the new predictions
accuracy_balanced = accuracy_score(y_test_discrete, y_pred)
print(f"\nModel Accuracy with balanced class weights: {accuracy_balanced:.4f}")

# Generate classification report with the new predictions
report_balanced = classification_report(y_test_discrete, y_pred)
print("\nClassification Report with balanced class weights:\n", report_balanced)


Model Accuracy with balanced class weights: 0.5676

Classification Report with balanced class weights:
               precision    recall  f1-score   support

           1       0.20      0.00      0.01       212
           2       0.00      0.00      0.00        70
           3       0.00      0.00      0.00       157
           4       0.41      0.02      0.04       376
           5       0.57      0.99      0.72      1072

    accuracy                           0.57      1887
   macro avg       0.24      0.20      0.15      1887
weighted avg       0.43      0.57      0.42      1887



/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


**Reasoning**:
The previous classification report still shows poor performance for minority classes and `UndefinedMetricWarning`, indicating that `class_weight='balanced'` alone is not sufficient. To further improve the model, I will perform hyperparameter tuning using `GridSearchCV` on the `RandomForestClassifier` with a focus on parameters like `n_estimators` and `max_depth`, using `f1_weighted` as the scoring metric to account for class imbalance.



In [22]:
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier

print("Starting hyperparameter tuning for RandomForestClassifier...")

# Define the parameter grid to search
param_grid = {
    'n_estimators': [100, 200, 300], # Number of trees in the forest
    'max_depth': [10, 20, 30, None], # Maximum depth of the tree
    'min_samples_split': [2, 5], # Minimum number of samples required to split an internal node
    'min_samples_leaf': [1, 2] # Minimum number of samples required to be at a leaf node
}

# Instantiate the RandomForestClassifier with balanced class weights
rf_classifier = RandomForestClassifier(random_state=42, class_weight='balanced')

# Instantiate GridSearchCV
# Using 'f1_weighted' as the scoring metric to handle class imbalance
grid_search = GridSearchCV(
    estimator=rf_classifier,
    param_grid=param_grid,
    scoring='f1_weighted',
    cv=3, # Using 3-fold cross-validation for reasonable computation time
    n_jobs=-1, # Use all available cores
    verbose=2 # Verbose output for progress
)

# Fit GridSearchCV to the training data with discrete target
grid_search.fit(X_train, y_train_discrete)

print("\nHyperparameter tuning complete.")
print(f"Best parameters found: {grid_search.best_params_}")
print(f"Best F1-weighted score: {grid_search.best_score_:.4f}")

# Store the best model
best_model = grid_search.best_estimator_
print("Best model stored for further evaluation.")

Starting hyperparameter tuning for RandomForestClassifier...
Fitting 3 folds for each of 48 candidates, totalling 144 fits

Hyperparameter tuning complete.
Best parameters found: {'max_depth': 20, 'min_samples_leaf': 2, 'min_samples_split': 2, 'n_estimators': 100}
Best F1-weighted score: 0.4158
Best model stored for further evaluation.


**Reasoning**:
After hyperparameter tuning, I will use the `best_model` obtained from `GridSearchCV` to make predictions on the test set (`X_test`) to evaluate its performance with optimized parameters.



In [23]:
y_pred_tuned = best_model.predict(X_test)

print("Model predictions on X_test generated using the best tuned model.")
print("First 5 predictions from tuned model:")
print(y_pred_tuned[:5])

Model predictions on X_test generated using the best tuned model.
First 5 predictions from tuned model:
[5 5 5 5 5]


**Reasoning**:
Now that predictions have been generated from the best-tuned model, I will calculate the accuracy score and a classification report using `y_test_discrete` and `y_pred_tuned` to evaluate its performance, as specified in the instructions.



In [24]:
from sklearn.metrics import accuracy_score, classification_report

# Calculate accuracy with the new predictions from the tuned model
accuracy_tuned = accuracy_score(y_test_discrete, y_pred_tuned)
print(f"\nModel Accuracy with tuned parameters: {accuracy_tuned:.4f}")

# Generate classification report with the new predictions from the tuned model
report_tuned = classification_report(y_test_discrete, y_pred_tuned)
print("\nClassification Report with tuned parameters:\n", report_tuned)


Model Accuracy with tuned parameters: 0.4865

Classification Report with tuned parameters:
               precision    recall  f1-score   support

           1       0.10      0.06      0.07       212
           2       0.00      0.00      0.00        70
           3       0.02      0.01      0.01       157
           4       0.20      0.10      0.13       376
           5       0.57      0.81      0.67      1072

    accuracy                           0.49      1887
   macro avg       0.18      0.19      0.18      1887
weighted avg       0.38      0.49      0.41      1887



## Análisis de Resultados y Conclusiones

### Subtask:
Analizar los resultados del modelo, identificar patrones, y sacar conclusiones. Esto puede incluir visualizaciones para comunicar los hallazgos.


### Análsis de Resultados y Conclusiones del Modelo Tuned

El `RandomForestClassifier`, después de la sintonización de hiperparámetros con `GridSearchCV` y el manejo de pesos de clase balanceados, mostró el siguiente rendimiento en el conjunto de prueba:

**Rendimiento General:**
*   **Precisión (Accuracy):** El modelo alcanzó una precisión de **0.4865**, lo que significa que aproximadamente el 48.65% de las predicciones del modelo fueron correctas. Este valor es moderado y sugiere que el modelo no es altamente preciso en todas las predicciones.
*   **F1-Score Ponderado (Weighted Avg F1-score):** El F1-score ponderado fue de **0.41**, lo cual es relativamente bajo. El F1-score es una media armónica de precisión y recall, y el valor ponderado considera el desequilibrio de clases, por lo que un valor bajo indica un rendimiento deficiente en general, especialmente en la capacidad de equilibrar la precisión y el recall en todas las clases.

**Rendimiento por Clase (Clasificación Report):**
```
               precision    recall  f1-score   support

           1       0.10      0.06      0.07       212
           2       0.00      0.00      0.00        70
           3       0.02      0.01      0.01       157
           4       0.20      0.10      0.13       376
           5       0.57      0.81      0.67      1072

    accuracy                           0.49      1887
   macro avg       0.18      0.19      0.18      1887
weighted avg       0.38      0.49      0.41      1887
```

*   **Clases Minoritarias (1, 2, 3, 4):** El rendimiento del modelo en las clases minoritarias es muy deficiente. Para las clases 2 y 3, la precisión, el recall y el F1-score son prácticamente cero (o se muestran como 0.00 debido a `zero_division` warnings). Esto significa que el modelo no fue capaz de predecir correctamente ninguna instancia de estas clases, o lo hizo en un número insignificante. Para las clases 1 y 4, los valores de precisión, recall y F1-score son extremadamente bajos (e.g., F1-score de 0.07 para clase 1, y 0.13 para clase 4). Esto confirma que el modelo está luchando significativamente para identificar estas categorías.
*   **Clase Mayoritaria (5):** Como se observa en el informe de clasificación, la clase 5, que es la clase mayoritaria (1072 instancias en el conjunto de prueba), domina las predicciones. Para esta clase, el modelo tiene un recall muy alto (0.81) y un F1-score decente (0.67), con una precisión de 0.57. Esto indica que el modelo tiende a predecir la clase 5 con mucha frecuencia, lo que infla la precisión general (accuracy) pero a expensas de las clases minoritarias.

**Conclusiones:**
El modelo `RandomForestClassifier`, a pesar de usar `class_weight='balanced'` y de la optimización de hiperparámetros, sigue sufriendo de un **fuerte sesgo hacia la clase mayoritaria (Calificación 5)**. La `UndefinedMetricWarning` en los pasos anteriores y los bajos valores de precisión y recall para las calificaciones 1, 2, 3 y 4 confirman que el desequilibrio de clases es un problema crítico que no se ha resuelto satisfactoriamente. El modelo prácticamente ignora las clases minoritarias, lo que lo hace ineficaz para predecir un rango completo de calificaciones.

**Propuestas para Futuros Pasos:**
Para mejorar el rendimiento del modelo en la predicción de todas las categorías de calificación, se deben considerar los siguientes enfoques:

1.  **Técnicas Avanzadas de Re-muestreo:** Explorar y aplicar técnicas más sofisticadas para manejar el desequilibrio de clases, como SMOTE (Synthetic Minority Over-sampling Technique), ADASYN (Adaptive Synthetic Sampling), o combinaciones de oversampling y undersampling. Estas técnicas pueden generar muestras sintéticas de las clases minoritarias o reducir las de las mayoritarias para crear un conjunto de datos más balanceado.
2.  **Explorar Diferentes Algoritmos:** Probar otros algoritmos de clasificación que sean inherentemente más robustos ante el desequilibrio de clases o que ofrezcan mecanismos de ponderación de clases más avanzados. Esto podría incluir:
    *   **LightGBM/XGBoost:** Modelos de boosting que pueden ser configurados para manejar el desequilibrio de clases.
    *   **Support Vector Machines (SVM):** Con ajuste de pesos de clase.
    *   **Redes Neuronales:** Con funciones de pérdida personalizadas o técnicas de muestreo incorporadas.
3.  **Recopilación de Más Datos:** Si es posible, buscar y recopilar más datos para las clases subrepresentadas (calificaciones 1, 2, 3 y 4). Esta es a menudo la solución más efectiva, aunque no siempre factible.
4.  **Ajuste de Hiperparámetros Más Profundo:** Realizar una búsqueda más exhaustiva de hiperparámetros, posiblemente utilizando `RandomizedSearchCV` antes de `GridSearchCV` para explorar un espacio de parámetros más amplio, o un rango de valores más fino para los parámetros clave.

## Generación de Predicciones y Entrega

### Subtask:
Generar las predicciones finales con el modelo entrenado y documentar el proceso.


## Summary:

### Data Analysis Key Findings

*   The initial dataset, comprising four CSV files, was successfully loaded and concatenated into a single DataFrame containing 9435 rows and 12 columns, with no missing values.
*   The 'Fecha de Compra' column was successfully transformed into several new temporal features, including year, month, day, day of week, day of year, and 'Dias\_Desde\_Primera\_Compra'.
*   Categorical features were one-hot encoded, expanding the DataFrame to 103 columns, and numerical features, including the original 'Precio', 'Costo de envío', 'Calificación', 'Cantidad de cuotas', 'lat', 'lon', and the newly engineered temporal features, were scaled using `StandardScaler`.
*   The target variable, 'Calificación', was initially scaled to a continuous range. For classification, it was inverse-transformed, rounded, and clipped back to discrete integer values ranging from 1 to 5.
*   An initial `RandomForestClassifier` model achieved an accuracy of approximately 56%, but showed severe limitations in predicting minority classes, with metrics like F1-score being 0.00 for several rating categories (e.g., 1, 2, 3).
*   Applying `class_weight='balanced'` and performing hyperparameter tuning with `GridSearchCV` (optimizing for `f1_weighted`) led to a best overall F1-weighted score of 0.4158 during cross-validation and a test accuracy of 48.65%.
*   Despite tuning, the model continued to exhibit a strong bias towards the majority class (Rating 5), achieving a high recall of 0.81 and an F1-score of 0.67 for this class. Conversely, it performed extremely poorly on minority classes, with F1-scores ranging from 0.00 (for ratings 2 and 3) to 0.13 (for rating 4), indicating that these classes were largely ignored.

### Insights or Next Steps

*   The current model is significantly hindered by severe class imbalance, leading to ineffective predictions for minority rating categories. Future efforts should prioritize advanced re-sampling techniques (e.g., SMOTE, ADASYN) to create a more balanced training dataset.
*   Explore alternative machine learning algorithms that are inherently more robust to class imbalance or offer more sophisticated class weighting mechanisms, such as LightGBM, XGBoost, Support Vector Machines, or Neural Networks, and conduct a deeper hyperparameter search for these models.
